### PR TITLE
Support for null/undefined children

### PR DESCRIPTION
**Pull request by nunocastromartins**
Changed the patching algorithm to support `null`/`undefined` children as I described in #226.

This should make the core algorithm slightly more resilient, but maybe changing the `h` function would be simpler (as was originally proposed). :-)

Tell me your thoughts. Maybe this needs more tests? Lazily copied the one @staltz wrote.

### CODE REVIEW

File: test/core.js [hunk 1]
```diff
@@ -230,6 +230,14 @@ describe('snabbdom', function() {
...
       assert.equal(elm.src, undefined);
     });
+    it('supports some null children', function() {
+      var vnode1 = h('span', [h('span', 'hi'), h('span', 'snabbdom')]);
+      var vnode2 = h('span', [h('span', 'hi'), null, h('span', 'snabbdom')]);
+      elm = patch(vnode0, vnode1).elm;
+      assert.equal(elm.children.length, 2);
+      elm = patch(vnode1, vnode2).elm;
+      assert.deepEqual(map(inner, elm.children), ['hi', 'snabbdom']);
+    });
```
Comment (mightyiam) [inline test/core.js]: I think there should be a test for `undefined` as well.

**Comment by staltz** [comment]
Might be a good opportunity to replace all `isDef` and `isUndef` cases.

**Comment by nunocastromartins** [comment]
> I think there should be a test for `undefined` as well.

I agree! I'll commit something in a bit.

> Might be a good opportunity to replace all `isDef` and `isUndef` cases.

Just so we're on the same page, you mean replace all occurrences of `isDef` and `isUndef` with `!= null` and `== null` checks? If so, then I agree.

**Comment by staltz** [comment]
We need @paldepind's opinion about that.

**Comment by paldepind** [comment]
I think this is a good idea and I think this PR looks like a good approach 👍

I think adding a few more to the tests may be nice. For instance in the edge case where all child values are `null`/undefined`. And maybe also a case where an element is toggled on and off. I.e. first it is a real vnode, then `undefined` and then a real vnode again.

**Comment by staltz** [comment]
By the way `foo === void 0` might be better than `foo === undefined`. A bit shorter.

**Comment by nunocastromartins** [comment]
> I think adding a few more to the tests may be nice. For instance in the edge case where all child values are `null`/`undefined`. And maybe also a case where an element is toggled on and off. I.e. first it is a real vnode, then `undefined` and then a real vnode again.

Agree.

> I personally prefer `foo === undefined` when just `undefined` is expected and to only use `foo == null` when both `undefined` and `null` are expected.

Just to be clear, the only place where `null` is expected is as a child, correct?

> By the way `foo === void 0` might be better than `foo === undefined`. A bit shorter.

I don't see a problem with this. `void 0` should also be safer, though I'm not sure that matters nowadays.

**Comment by nunocastromartins** [comment]
> And maybe also a case where an element is toggled on and off. I.e. first it is a real vnode, then `undefined` and then a real vnode again.

@paldepind I _think_ this is being tested. Tell me if you'd like some other test.

Regarding the change from `undefined` to `void 0`, I abstained from doing so in this latest commit; if you'd like me to change it, tell me so. :-)

**Comment by paldepind** [comment]
I'm fairly certain that Uglify transforms `undefined` to `void 0`. But with regards to code size I think keeping `isUndef` is best since the function call can be made really short.

**Comment by nunocastromartins** [comment]
So you'd rather keep the `isDef`/`isUndef` checks?

**Comment by staltz** [comment]
@nunocastromartins could you add back the `isDef`/`isUndef`? We could discuss that in another issue/PR, but it would be nice to get this PR merged for the purpose of supporting null children.

**Comment by nunocastromartins** [comment]
@staltz, readded `isDef`/`isUndef` and rebased PR.

**Comment by paldepind** [comment]
Thank you @nunocastromartins :+1:
